### PR TITLE
Added billboard text annotations

### PR DESCRIPTION
**Pull request by mcjohnalds**
Fixes issue #147 

Example:
```
displaz autzen.laz &
displaz -annotate "exploded thing" 637156.720 850102.120 485.140 &
displaz -annotate "garbage" 636837.230 851533.400 428.570 &
displaz -annotate "delicious eggplant 🍆" 638043.230 851172.570 421.130
```
Result:
annotations

- Annotations stay at a fixed size regardless of distance
- Annotations always face the camera
- Annotations are not obscured by scene geometry
- You can toggle displaying annotations by clicking View->Draw Annotations.

I got the automatic switching between fixed-size and geometric scaling based on distance from camera working (as described in the comments) but I couldn't work out the maths required for a smooth transition between the two so I left that out for now. I'll try and get that working later.

### CODE REVIEW

File: shaders/billboard.glsl [hunk 1]
```diff
@@ -0,0 +1,47 @@
+#version 150
+// Copyright 2015, Christopher J. Foster and the other displaz contributors.
+// Use of this code is governed by the BSD-style license found in LICENSE.txt
+
+// Draws billboards, which are textured quads that always face the camera,
```
Comment (nigels-com) [inline shaders/billboard.glsl]: I don't think these are billboards in the usual usage of the term.  I suggest renaming the shader `annotations.glsl`

File: src/render/Billboard.cpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
+// Copyright 2015, Christopher J. Foster and the other displaz contributors.
+// Use of this code is governed by the BSD-style license found in LICENSE.txt
+
+#include "Billboard.h"
```
Comment (nigels-com) [inline src/render/Billboard.cpp]: SImilarly for `Billboard.cpp` and `Billboard.h`, rename to `Annotation.h/cpp`

**Comment by nigels-com** [comment]
Looks good to me.  I'd suggest replacing the billboarding terminology with annotation.  Billboarding is usually for approximating 3D geometry with pre-rendered textures.

**Comment by mcjohnalds** [comment]
Everything's been renamed from billboard to annotation.

File: shaders/annotation.glsl [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+uniform mat4 modelViewMatrix;
+uniform mat4 projectionMatrix;
+uniform mat4 modelViewProjectionMatrix;
+uniform vec2 annotationSize; // In pixels
+uniform vec2 viewportSize; // In pixels
+
+//------------------------------------------------------------------------------
+#if defined(VERTEX_SHADER)
+
+in vec2 position;
```
Comment (c42f) [inline shaders/annotation.glsl]: A couple of comments about the coordinate system of `position` would be helpful here or down below: `makeVAO`, `Annotation::draw()` and the shader work together in a tightly coupled way, and it's just a little tricky to figure out what's going on.

**Comment by mcjohnalds** [comment]
Not sure why the dates on my previous commits got garbled when I pushed but there we go.

**Comment by c42f** [comment]
I think this is almost ready to merge; as a first go at an annotation system it's already going to be very useful.  I noticed you're missing one small thing which is to call `restartRender()` inside `addAnnotation()` to force the labels to show up immediately.  (displaz has an event-based OpenGL loop, and doesn't draw anything when the view isn't changing.)

Also, I'd like to explore the command line API just a little more so we can have something forward compatible for scripts to rely on.  Eg, what if we later want a feature to allow people to be replace existing annotations on a geometry?  It would be nice if `-annotate` can expand to cover that use case (and potentially others).

I don't know what's nicest here, just throwing out a few ideas:

```
displaz -annotation "text=$text,parent_geometry=$label" $x $y $z

# Hmm, how does this work out if we wanted other annotations, eg an arrow marker?
displaz -annotation "style=arrow,parent_geometry=$label" $x $y $z

# Or if we wanted to remove a text annotation
displaz -annotation "remove=true,text=$text,parent_geometry=$label" $x $y $z
```
Hmm, are x,y,z are irrelevant in the case of removal?  How should the user refer to the identity of an annotation?  An optional label for an annotation (distinct from the text) might make sense, in analogy to the geometry labelling system which I think is working quite well.

Thoughts?

**Comment by mcjohnalds** [comment]
If you anticipate displaz's API growing a lot it might be worth doing what git does with its subcommands e.g,

```
displaz add [-label <label>] <path>
displaz unload <label>
displaz annotation [-label <label>] [-arrow] <text> <x> <y> <z>
...
```

That way you can overload commands like annotation with lots of options while keeping a straightforward interface. You could leave the existing interface deprecated and undocumented. Ignoring that, if we were to give annotations their own labels, then `-label <label>` and `-unload <label>` could be made to work with annotations, removing the need for parent geometry. E.g,

```
displaz -annotation "a string" 12.3 45.6 78.9 # "annotation1" label generated
displaz -label "foo" -annotation "another string" 98.7 65.4 32.1 # "foo" label generated
displaz -unload "annotation1" # first annotation removed
```

I'll wait until we finalise the API design before I mess with the signals and slots code to call `restartRenderer()`. In fact you might want to leave this PR on the backburner until the API is figured out.

Side note: If you want a simple arrow to point to things then using any of 🡳🡻🢃▼🠷 works fine.

**Comment by c42f** [comment]
Good points, I like the look of the subcommands version.

The displaz command line has slowly grown and at this stage it's definitely getting a bit inconsistent and unwieldy.  My rough plan is to keep growing it organically a while longer and then revisit it in a big refactor.  For the moment I'm putting off the refactor because I want to have a bigger list of operations before trying to generalize to something nicer.  I do wonder whether the definitive API should be in a documented socket protocol (rather than the mess I have now :-)), with the command line just a convenient wrapper around that.

Anyway, that's future stuff.  For the moment, `-label` will do nicely for labelling annotations, but I do think attaching them to parent geometry would be a very useful feature.  (Related - one thing I've been considering for a long time is having geometry lifetime and grouping managed in a tree structure, but I just haven't had the time to do this properly.)

Perhaps the right thing is to just let users rely on naming conventions for now, rather than having the ownership semantics of annotations vs geometry being strangely inconsistent.  This would be quite usable from a script:

```
displaz -label blah some/geometry.ply
displaz -label "blah:foo" -annotation "foo" 1 2 3 # pure label naming convention to indicate grouping
displaz -label "blah:foo2" -annotation "foo2" 4 5 6
displaz -label "blah:foo" -annotation "foo3" 7 8 9  # Replaces "blah:foo"

displaz -unload "blah*" # remove blah and its annotations
```

Then later on, if tree-based grouping of geometries is added, annotations could just be part of that.

**Comment by mcjohnalds** [comment]
To clarify, if I did

```
displaz -label blah some/geometry.ply
displaz -label "blah:foo" -annotation "foo" 1 2 3
displaz -unload blah
```

then the "blah:foo" annotation should still be there?

**Comment by c42f** [comment]
Cprrect.  `-unload` is a bit of a special case in that it takes a shell glob-style pattern.  So `blah*` matches `blah:foo`, but `blah` does not, according to the usual shell globbing rules.

**Comment by mcjohnalds** [comment]
-annotation should now work as you've described

**Comment by c42f** [comment]
This is a great addition, and I'm really sorry to have taken so long to get back to it.  I'm just going to merge this now!

**Pull request merged**